collectors/modality-probe-debug-collector/src/lib.rs

```diff
@@ -0,0 +1,1345 @@
+        let write_seqn_low_addr = hist_addr + write_seqn_low_offset();
+        let overwrite_seqn_high_addr = hist_addr + overwrite_seqn_high_offset();
+        let overwrite_seqn_low_addr = hist_addr + overwrite_seqn_low_offset();
+        let priority_field_addr = hist_addr + overwrite_priority_offset();
+
+        let priority_mem_accessor = mem_accessor.clone();
+        let mut clocks = Vec::new();
+        // Merge self clock set to 0
+        Self::merge_clock(
+            &mut clocks,
+            LogicalClock {
+                id,
+                epoch: ProbeEpoch(0),
+                ticks: ProbeTicks(0),
+            },
+        );
+        Ok(Self {
+            seq_num: 0,
+            reader: FencedReader::new(
+                MemorySnapper {
+                    mem_accessor,
+                    storage_addr: buf_addr,
+                    write_seqn_high_addr,
+                    write_seqn_low_addr,
+                    overwrite_seqn_high_addr,
+                    overwrite_seqn_low_addr,
+                },
+                buf_cap,
+            ),
+            rbuf: Vec::new(),
+            clocks,
+            priority_writer: PriorityWriter {
+                mem_accessor: priority_mem_accessor,
+                priority_field_addr,
+            },
+        })
+    }
+
+    fn find_probe(
+        probe_addr: &ProbeAddr,
+        mem_accessor: Rc<RefCell<dyn MemoryAccessor>>,
+    ) -> Result<Word, Error> {
+        match *probe_addr {
+            ProbeAddr::Addr(addr) => {
+                let padded_probe_address = Self::scan_guard_bytes(addr, mem_accessor.clone())?;
+                // Ensure probe has correct fingerprint
+                let fingerprint = mem_accessor.borrow_mut().read_32(padded_probe_address)?;
+                if fingerprint != ModalityProbe::STRUCT_FINGERPRINT {
+                    Err(Error::NoProbeAtAddress(addr.into()))
+                } else {
+                    Ok(padded_probe_address)
+                }
+            }
+            // Read address from pointer
+            ProbeAddr::PtrAddr(ptr_addr) => {
+                let deref_addr = mem_accessor.borrow_mut().read_word(ptr_addr)?;
+                let padded_probe_address =
+                    Self::scan_guard_bytes(deref_addr, mem_accessor.clone())?;
+
+                // Ensure probe has correct fingerprint
+                let fingerprint = mem_accessor.borrow_mut().read_32(padded_probe_address)?;
+                if fingerprint != ModalityProbe::STRUCT_FINGERPRINT {
+                    Err(Error::InvalidProbePointer(ptr_addr.into()))
+                } else {
+                    Ok(padded_probe_address)
+                }
+            }
+        }
+    }
+
+    fn scan_guard_bytes(
+        mut addr: Word,
+        mem_accessor: Rc<RefCell<dyn MemoryAccessor>>,
+    ) -> Result<Word, Error> {
+        let mut mem = mem_accessor.borrow_mut();
+        loop {
```

Comment:
This loop ought to be bounded by a constant.  One candidate off the top of my head is the alignment-size of the ModalityProbe (or a small multiple thereof).